Python/pythonrun.c

```diff
@@ -110,24 +112,31 @@ PyRun_InteractiveLoopFlags(FILE *fp, const char *filename_str, PyCompilerFlags *
         _PySys_SetObjectId(&PyId_ps2, v = PyUnicode_FromString("... "));
         Py_XDECREF(v);
     }
-    err = -1;
-    for (;;) {
-        ret = PyRun_InteractiveOneObject(fp, filename, flags);
+    while (ret != E_EOF) {
+        ret = PyRun_InteractiveOneObjectEx(fp, filename, flags);
+        if (ret == -1 && PyErr_Occurred()) {
+            /* Prevent an endless loop after multiple consecutive MemoryErrors
+             * while still allowing an interactive command to fail with a
+             * MemoryError. */
+            if (PyErr_ExceptionMatches(PyExc_MemoryError)) {
+                if (++nomem_count > 16) {
+                    Py_FatalError("Cannot recover from MemoryErrors.");
```

Comment:
Please don't call Py_FatalError(), but return with an error code.

Reply:
When returning -1 and exiting at the first MemoryError and there is indeed no memory left upon returning -1, the error message does not indicate that there has been a MemoryError:
```
$ ./python -q
>>> from _testcapi import set_nomemory
>>> set_nomemory(0)
Error in atexit._run_exitfuncs:
Error in atexit._run_exitfuncs:
$
```

Reply:
In real life Py_FatalError() would also fail to print the error msg since printf() may rely on malloc(). And returning -1 is consistent with the behavior of the other PyRun_*() functions.

I will change PyRun_InteractiveLoopFlags() to return -1 as you have suggested :-)